tests/test-cross-context-dispatch/ccd-tests/src/test/java/org/eclipse/jetty/tests/redispatch/RedispatchPlansTests.java

```diff
@@ -0,0 +1,176 @@
+            .path(requestStep.getRequestPath())
+            .send();
+        String responseDetails = toResponseDetails(response);
+        assertThat(responseDetails, response.getStatus(), is(HttpStatus.OK_200));
+
+        Properties responseProps = new Properties();
+        try (StringReader stringReader = new StringReader(response.getContentAsString()))
+        {
+            responseProps.load(stringReader);
+        }
+
+        dumpProperties(responseProps);
+
+        int expectedEventCount = dispatchPlan.getExpectedEvents().size();
+        assertThat(responseProps.getProperty("dispatchPlan.events.count"), is(Integer.toString(expectedEventCount)));
+        for (int i = 0; i < expectedEventCount; i++)
+        {
+            assertThat("id[" + dispatchPlan.id() + "] event[" + i + "]", responseProps.getProperty("dispatchPlan.event[" + i + "]"), is(dispatchPlan.getExpectedEvents().get(i)));
+        }
+
+        if (dispatchPlan.getExpectedContentType() != null)
+        {
+            assertThat("Expected ContentType", response.getHeaders().get(HttpHeader.CONTENT_TYPE), is(dispatchPlan.getExpectedContentType()));
+        }
+
+        for (Property expectedProperty : dispatchPlan.getExpectedProperties())
+        {
+            assertProperty(dispatchPlan.id(), responseProps, expectedProperty.getName(), is(expectedProperty.getValue()));
+        }
+
+        // Ensure that all seen session ids are the same.
+        if (dispatchPlan.isExpectedSessionIds())
+        {
+            // Verify that Request Attributes for Session.id are in agreement
+            List<String> attrNames = responseProps.keySet().stream()
+                .map(Object::toString)
+                .filter((name) -> name.startsWith("req.attr[session["))
+                .toList();
+
+            if (attrNames.size() > 1)
+            {
+                String expectedId = responseProps.getProperty(attrNames.get(0));
+                for (String name : attrNames)
+                {
+                    assertEquals(expectedId, responseProps.getProperty(name));
+                }
+            }
+
+            // Verify that Context Attributes for Session.id are in agreement
+            // And that all ids have had their .commit() and .release() methods called.
+            Path sessionLog = jettyBase.jettyBase.resolve("work/session.log");
+            assertTrue(Files.isRegularFile(sessionLog), "Missing " + sessionLog);
+
+            List<String> logEntries = Files.readAllLines(sessionLog);
+            List<String> newSessions = logEntries.stream()
+                .filter(line -> line.contains("SessionCache.event.newSession()"))
+                .map(line -> line.substring(line.indexOf("=") + 1))
+                .toList();
+            // we should have the commit() and release() for each new Session.
+            for (String sessionId : newSessions)
+            {
+                assertThat(logEntries, hasItem("SessionCache.event.commit()=" + sessionId));
```

Comment:
I think there's a race condition here - I think that the log might not be finished being written by the time you're looking at the output, as I see tests of these session contents failing randomly, even though the log contains the correct information. So you need to use `Awaitility` to wait a bit. Have a look for example at https://github.com/jetty/jetty.project/blob/jetty-12.0.x/jetty-ee9/jetty-ee9-tests/jetty-ee9-test-sessions/jetty-ee9-test-sessions-common/src/test/java/org/eclipse/jetty/ee9/session/CreationTest.java#L157

Reply:
In this test suite we cannot dig down into a class in one of the EE# environments (we don't have access to it).
We don't even have access to that JVM for waiting on the `Server` instance, as it's a forked JVM.
That means we have no test we can give `Awaitility`.

The best we can do is setup graceful shutdown and wait on the JVM closing via the `JettyHomeTester`.

Reply:
All you need to do is to wait until the `session.log` file is no longer being written to.

Reply:
> All you need to do is to wait until the `session.log` file is no longer being written to.

That's not a state check you can do from a different process (pid).
Keep in mind we are dealing with a forked JVM.

Waiting for the forked JVM to exit seems like the only event we can use.